multi: accept memo note when opening channel

shaurya947 opened the pull request:
## Change Description
Fixes #7639 by allowing a note ("memo") to go along with channel creation when using the openchannel CLI/RPC/JSON command. We accept a string with max length of 500 (command is rejected if length is exceeded) and persist it in a hex-encoded format as part of the TLV stream. When channels are queried, we retrieve the value and decode it before passing the info to the caller. We also add a new itest to verify both the happy and unhappy path.

## Steps to Test
When using openchannel through CLI/RPC/JSON, pass in a memo note using the `memo` argument. If length is valid (<= 500), it should persist and be retrievable in listchannels call. If length is invalid, the command should be rejected with an appropriate error message.

## Pull Request Checklist
### Testing
- [ ] Your PR passes all CI checks.
- [X] Tests covering the positive and negative (error paths) are included.
- [X] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [X] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [X] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [X] Any new logging statements use an appropriate subsystem and logging level.
- [X]  [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our  [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

shaurya947 commented:
Looks like `TestFetchHistoricalChannel` is failing inside `db_test.go` and I'm not quite sure why. It fails when deep-comparing `channel` with `histChannel`:
https://github.com/shaurya947/lnd/blob/cd2bea92db0dc1b45b84d69d15e65ffba8d674eb/channeldb/db_test.go#L693

The test seems to pass if I attach a `Memo` here:
https://github.com/shaurya947/lnd/blob/cd2bea92db0dc1b45b84d69d15e65ffba8d674eb/channeldb/channel_test.go#L313

But that shouldn't be required right? Am I doing something wrong with the TLV stream which messes up (de)serialization when the `Memo` field is not present?

guggero commented on rpcserver.go:
```diff
@@ -2169,6 +2169,13 @@ func (r *rpcServer) parseOpenChannelReq(in *lnrpc.OpenChannelRequest,
+			"exceeds 500", in.Memo, len(in.Memo))
+	}
+	memoHex := make([]byte, hex.EncodedLen(len(in.Memo)))
+	hex.Encode(memoHex, []byte(in.Memo))
```
Why do we use up extra space by hex encoding the memo? Can't we just use `[]byte(in.Memo)` directly in the channel?

shaurya947 replied:
I'm honestly not sure what the long-term benefits of hex-encoding would be. There was a bit of discussion [here](https://github.com/lightningnetwork/lnd/issues/7639#issuecomment-1533762154). You think we should just ditch the hex-encoding/decoding?

guggero replied:
I don't see any benefit with the hex encoding either. Since we're not interpreting the data on the DB level and are only reading it length-delimited, there's also no potential of special characters interfering with the storage logic (which it might do if we were using SQL). So I'd say we can ditch the hex encoding.

guggero commented on channeldb/channel.go:
```diff
@@ -3642,6 +3651,7 @@ func putChanInfo(chanBucket kvdb.RwBucket, channel *OpenChannel) error {
 			initialRemoteBalanceType, &remoteBalance,
 		),
 		MakeScidRecord(realScidType, &channel.confirmedScid),
+		tlv.MakePrimitiveRecord(channelMemoType, &channel.Memo),
```
Even though I like small commits that are easy to review, this PR is a bit on the "too small" side IMO. So things that are very closely related (as for example the encoding and decoding of the same record) should be grouped together, to make it easier to make sure nothing is forgotten. Especially if it's just a couple of lines and not hundreds of them. So I would combine this and the next commit.

guggero commented on channeldb/channel.go:
```diff
@@ -3869,6 +3869,7 @@ func fetchChanInfo(chanBucket kvdb.RBucket, channel *OpenChannel) error {
 			initialRemoteBalanceType, &remoteBalance,
 		),
 		MakeScidRecord(realScidType, &channel.confirmedScid),
+		tlv.MakePrimitiveRecord(channelMemoType, &channel.Memo),
```
I was able to track down the weird crash in the unit test to this. I think you need to use an extra variable here (e.g. `var memo []byte`, decode into that (`tlv.MakePrimitiveRecord(channelMemoType, &memo),` and then check if something was decoded:
```
if len(memo) > 0 {
    channel.Memo = memo
}
```

This seems to be a weird `nil` slice vs. empty slice thing in the deep comparison...

shaurya947 replied:
Thanks for looking into that! I will try your suggestion :slightly_smiling_face:

guggero commented on rpcserver.go:
```diff
@@ -4191,6 +4191,10 @@ func createRPCOpenChannel(r *rpcServer, dbChannel *channeldb.OpenChannel,
 
+	// Hex-decode the channel memo. Ignore the error (memo will be empty).
+	memo := make([]byte, hex.DecodedLen(len(dbChannel.Memo)))
+	_, _ = hex.Decode(memo, dbChannel.Memo)
```
Same comment here re unnecessary hex encoding.

guggero commented on itest/list_on_test.go:
```diff
@@ -535,4 +535,8 @@ var allTestCases = []*lntest.TestCase{
 		TestFunc: testCustomFeatures,
 	},
+	{
+		Name:     "channel memo during creation",
```
Could this perhaps be folded in with an existing test? Just to reduce the number of test cases to run, as each of them have some overhead (and we're well into the full hour or more runtime on those now).

shaurya947 replied:
I suppose I could stuff this into `testBasicChannelCreationAndUpdates` here:
https://github.com/lightningnetwork/lnd/blob/fd9adaf6ceb3649c07bbb4982bd60dd632e8cda0/itest/lnd_open_channel_test.go#L379

What do you think?

guggero replied:
Sounds good, yes!

guggero commented on itest/lnd_open_channel_test.go:
```diff
@@ -357,6 +358,38 @@ func testBasicChannelCreationAndUpdates(ht *lntest.HarnessTest) {
+// the memo is retrieved correctly when querying the created channel. We also
+// test that the 500 character limit is observed correctly.
+func testChannelMemoDuringCreation(ht *lntest.HarnessTest) {
+	// open a channel from Alice to Bob with a valid memo
```
nit: Inline comments should also be full sentences with capitalization and punctuation.

guggero reviewed (commented):
Very nice feature and well structured PR, thanks a lot!

shaurya947 commented:
Hey @guggero I updated the PR to address all your feedback:
- Ditch hex encoding
- Combine `dbchannel` changes into a single commit
- Combine `lnrpc` changes into a single commit
- Reuse existing itest

Also thanks so much for helping with the unit test failure. Your suggestion seems to have worked :slightly_smiling_face:

guggero reviewed (approved):
Just a couple of more nits, otherwise LGTM :tada:

guggero commented on itest/lnd_open_channel_test.go:
```diff
@@ -445,6 +463,14 @@ func runBasicChannelCreationAndUpdates(ht *lntest.HarnessTest,
+	ht.OpenChannelAssertErr(ht.Bob, ht.Alice, lntest.OpenChannelParams{
+		Amt:  funding.MaxBtcFundingAmount,
+		Memo: invalidMemo,
+	}, fmt.Errorf("provided memo (%s) is of length 501, exceeds 500",
```
nit: weird formatting. Maybe declare the `expectedErr` first, then pass it into the `OpenChannelAssertErr` function?

shaurya947 replied:
broke out the channel params as well to aid readability

shaurya947 commented:
Hey @guggero addressed all your feedback as well as fixed a silly mistake I made in the itest (used `ht.Alice`/`ht.Bob` instead of the `alice` and `bob` arguments)

shaurya947 commented:
Hey @guggero the windows itest check seems to fail because of a process lock on a file it seems? Is that expected?

Also any suggestions as to whom I could tag for a second review?

positiveblue reviewed (approved):
LGTM :rocket: 

This is such a quick win. I can see a lot of usage form external software that opens channels automatically for example. 

I think we should make the UX more coherent by also adding this field to the `PendingChannels` proto message but nothing blocking.

ffranr commented on lnrpc/lightning.proto:
```diff
@@ -1545,6 +1545,13 @@ message Channel {
 
     // This is the peer SCID alias.
     uint64 peer_scid_alias = 35 [jstype = JS_STRING];
+
+    /*
+    An optional note-to-self to go along with the channel containing some
+    useful information. This is only ever stored locally and in no way impacts
+    the channel's operation.
+    */
+    string memo = 36;
```
I think this commit message suggests that the commit does a lot more than what it actually does. Consider revising the commit message.

shaurya947 replied:
Good catch, updated!

ffranr commented on itest/lnd_open_channel_test.go:
```diff
@@ -373,16 +374,32 @@ func runBasicChannelCreationAndUpdates(ht *lntest.HarnessTest,
+	channel := ht.QueryChannelByChanPoint(alice, chanPoints[0])
+	require.Equal(ht, "bob is a good peer", channel.Memo)
+
+	// Bob shouldn't see the memo since it's for Alice only.
+	channel = ht.QueryChannelByChanPoint(bob, chanPoints[0])
+	require.Empty(ht, channel.Memo, "Memo is not empty")
```
:+1:

ffranr reviewed (commented):
I enjoyed walking through the commits, nice work!

shaurya947 commented on lnrpc/lightning.proto:
```diff
@@ -1545,6 +1545,13 @@ message Channel {
+    useful information. This is only ever stored locally and in no way impacts
+    the channel's operation.
+    */
+    string memo = 36;
```
Will address in a follow-up PR if that's ok. Want to make sure I can handle all cases around pendingopen/pendingforceclose/pendingwaitingclose properly and not inadvertently cause any more regressions in this PR.

shaurya947 commented:
Thanks for the review @positiveblue and @ffranr ! :slightly_smiling_face: 

I fixed the one commit message and created a constant for the magic number (along with some comments). I will defer the `PendingChannels` change for a follow-up PR if that's ok. I've made a branch for it, I just want to make sure I can handle all the different pending states properly and not cause any more inadvertent regressions in this PR.

shaurya947 commented:
Thank you folks! My first PR :smiling_face_with_tear:
